### PR TITLE
Admin login API

### PR DESCRIPTION
**Pull request by annq3012**
## Description
-Handle API from Portal then save information of user into database
## Reference Requirements
- Product Backlog ID: PBI-1
- Task ID: #1.3

## Migrations
YES

## Impacted Areas in Application
- app/Http/Kernel.php
- composer.json
- config/auth.php
- config/define.php
- config/portal.php
- public/css/main.css
- resources/lang/en/home.php
- resources/lang/en/login.php
-  resources/lang/en/portal.php
- routes/web.php
- databaes/migration/2014_10_12_000000_create_users_table.php
## SQL query and `explain`
List of all queries and attach the result of each explaining query (in text, screenshot...)
select * from `users` where `id` = '8' and `users`.`deleted_at` is null limit 1
- Refer to [Using Explain with MySQL](https://viblo.asia/p/su-dung-explain-de-toi-uu-cau-lenh-mysql-BYjv44gmvxpV)
explain

### CODE REVIEW

File: app/Http/Requests/Backend/LoginRequest.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'email' => 'required',
```
Comment (at-anhngo) [inline app/Http/Requests/Backend/LoginRequest.php]: em nghĩ nên có giá trị email chỗ thuộc tính email

File: config/define.php [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?php 
+  return [
+    'datetime_format' => 'Y-m-d h:i:s',
+    'success' => 'successfully',
+  ];
```
Comment (at-anhngo) [inline config/define.php]: nên xuống dòng

File: config/portal.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+return [
+    'base_url_api' => 'http://172.16.110.17/api/v1',
+    
+    'end_point' => [
+        'login' => '/auth/login',
+        'logout' => '',
+    ]
+];
```
Comment (at-anhngo) [inline config/portal.php]: xuống dòng

File: resources/views/backend/home/index.blade.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                </div>
+            </div>
+        </div>
+    </section>
+    <!-- /.content -->
+  </div>
+  <!-- /.content-wrapper -->
+<!-- ./wrapper -->
+
+@endsection
```
Comment (at-anhngo) [inline resources/views/backend/home/index.blade.php]: xuống dòng cuối file

File: resources/views/backend/layouts/partials/js.blade.php [hunk 5]
```diff
@@ -23,6 +23,9 @@
 <!-- AdminLTE App -->
 <script src="{{ asset('bower_components/admin-lte/dist/js/adminlte.min.js') }}"></script>
 <!-- AdminLTE for demo purposes -->
+<<<<<<< Updated upstream
 <script src="{{ asset('bower_components/ckeditor/ckeditor.js') }}"></script>
-
 <script src="{{ asset('js/main.js') }}"></script>
+=======
+<script src="{{ asset('bower_components/admin-lte/dist/js/demo.js') }}"></script>
+>>>>>>> Stashed changes
```
Comment (at-anhngo) [inline resources/views/backend/layouts/partials/js.blade.php]: fix conflict

File: resources/views/backend/users/login.blade.php [hunk 6]
```diff
@@ -10,7 +10,13 @@
                 <div class="panel-body">
                     <form class="form-horizontal" method="POST" action="{{ route('login') }}">
                         {{ csrf_field() }}
-
+                        <div class="message">
+                            @if ($errors->has('message'))
+                                    <span class="help-block">
+                                        <strong>{{ $errors->first('message') }}</strong>
+                                    </span>
```
Comment (at-anhngo) [inline resources/views/backend/users/login.blade.php]: lỗi front end convention

**Review by ghost — CHANGES_REQUESTED** [review]

File: app/Http/Requests/Backend/LoginRequest.php (hunk 1 above)
Comment (at-annguyen) [reply app/Http/Requests/Backend/LoginRequest.php]: laravel have validation of email

**Comment by at-anhngo** [comment]
LGTM

**Review by ghost — APPROVED** [review]

**Comment by at-duyho** [comment]
- fix conflict and separate pull request. thank you.

File: app/Http/Controllers/Admin/LoginController.php [hunk 7]
```diff
@@ -26,4 +38,72 @@ public function showLoginForm()
...
+            # Catch errors from Portal
+            $portalResponse = json_decode($e->getResponse()->getBody()->getContents());
+            return redirect()
+                ->back()
+                ->withInput()
+                ->withErrors(['message' => trans('portal.messages.' . $portalResponse->meta->messages)]);
+        }
+        
+        # Check status API response
+        if ($portalResponse->meta->status == config('define.success')) {
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/LoginController.php]: - why don't you move it to try block?

File: app/Http/Middleware/Backend/AdminLoginMiddleware.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+     *
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if (Auth::check()) {
+            $user = Auth::user();
+            if ($user->role == User::$role['admin']) {
+                return $next($request);
+            } else {
```
Comment (at-duyho) [inline app/Http/Middleware/Backend/AdminLoginMiddleware.php]: - you has return, so you don't need else block.

File: app/Http/Requests/Backend/LoginRequest.php (hunk 1 above)
Comment (at-duyho) [inline app/Http/Requests/Backend/LoginRequest.php]: - add rule email

File: app/Model/User.php [hunk 9]
```diff
@@ -33,6 +33,8 @@ class User extends Authenticatable
         'team',
         'avatar_url',
         'role',
+        'access_token',
+        'expires_at'
```
Comment (at-duyho) [inline app/Model/User.php]: - change column and attribute user: expired_at

File: routes/web.php [hunk 10]
```diff
@@ -17,12 +17,12 @@
...
 Route::get('/login', 'Admin\LoginController@showLoginForm')->name('login');
-
-//Admin
-Route::group(['prefix' => 'admin', 'namespace' => 'Admin'], function () {
-    Route::resource('users', 'UserController', ['except' => ['create', 'store']]);
-    Route::resource('books', 'BookController');
+Route::post('/login', 'Admin\LoginController@login');
+Route::post('/logout', 'Admin\LoginController@logout')->name('logout');
+//Admin 
+Route::group(['prefix' => 'admin', 'namespace' => 'Admin', 'middleware' => 'admin'], function () {
```
Comment (at-duyho) [inline routes/web.php]: - you has route group admin on top file. please check it.

**Review by ghost — CHANGES_REQUESTED** [review]
check my comment

File: app/Http/Middleware/Backend/AdminLoginMiddleware.php (hunk 8 above)
Comment (at-annguyen) [reply app/Http/Middleware/Backend/AdminLoginMiddleware.php]: else block into LoginController at login method in try catch

**Comment by at-duyho** [comment]
- please, separate this pull request. thank you!

File: app/Http/Controllers/Admin/LoginController.php [hunk 11]
```diff
@@ -26,4 +38,75 @@ public function showLoginForm()
...
+            }
+        } catch (ServerException $e) {
+            # Catch errors from Portal
+            $portalResponse = json_decode($e->getResponse()->getBody()->getContents());
+            return redirect()
+                ->back()
+                ->withInput()
+                ->withErrors(['message' => trans('portal.messages.' . $portalResponse->meta->messages)]);
+        } catch (Exception $e) {
+            echo  $e->getMessage();
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/LoginController.php]: - custom page error.

File: app/Http/Middleware/Backend/AdminLoginMiddleware.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    public function handle($request, Closure $next)
+    {
+        if (Auth::check()) {
+            $user = Auth::user();
+            if ($user->role == User::ROOT_ADMIN) {
+                return $next($request);
+            }
+            return redirect('/')
+                ->withErrors(['message' => trans('portal.messages.not_an_admin')]);
+        } else {
```
Comment (at-duyho) [inline app/Http/Middleware/Backend/AdminLoginMiddleware.php]: - đã dùng return trong if rồi, thì khỏi cần phải dùng else. return về luôn.

File: app/Http/Controllers/Admin/LoginController.php [hunk 13]
```diff
@@ -26,4 +38,75 @@ public function showLoginForm()
...
+
+            # Check status API response
+            if ($portalResponse->meta->status == config('define.success')) {
+                $userResponse = $portalResponse->data->user;
+                # Collect user data from response
+                $teamName = $userResponse->teams[0]->name;
+                $date = date(config('define.datetime_format'), strtotime($userResponse->expires_at));
+                $userCondition = [
+                    'employee_code' => $userResponse->employee_code,
+                    'email' => $request->email,
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/LoginController.php]: what will happen when email is invalid AT format?

File: app/Http/Controllers/Admin/LoginController.php [hunk 14]
```diff
@@ -26,4 +38,75 @@ public function showLoginForm()
...
+        # Collect data form request
+        $data = $request->except('_token');
+        try {
+            # Try to call API to Portal
+            $client = new Client();
+            $portal = $client->post(config('portal.base_url_api') . config('portal.end_point.login'), ['form_params' => $data]);
+            $portalResponse = json_decode($portal->getBody()->getContents());
+
+            # Check status API response
+            if ($portalResponse->meta->status == config('define.success')) {
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/LoginController.php]: - you should check by status code

**Review by ghost — CHANGES_REQUESTED** [review]
check my comment.

File: app/Http/Controllers/Admin/LoginController.php [hunk 15]
```diff
@@ -26,4 +39,78 @@ public function showLoginForm()
...
+        # Collect data form request
+        $data = $request->except('_token');
+        try {
+            # Try to call API to Portal
+            $client = new Client();
+            $portal = $client->post(config('portal.base_url_api') . config('portal.end_point.login'), ['form_params' => $data]);
+            $portalResponse = json_decode($portal->getBody()->getContents());
+
+            # Check status API response
+            if ($portalResponse->meta->code == 200) {
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/LoginController.php]: - you should use Http::StatusOK

**Review by ghost — CHANGES_REQUESTED** [review]
check my comment

**Review by ghost — APPROVED** [review]
LGTM

**Pull request merged**